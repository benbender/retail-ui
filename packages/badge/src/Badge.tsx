import { ButtonIcon, ButtonIconVariant } from '@retail-ui/button-icon'
import { XSolid } from '@retail-ui/heroicons'
import { Theme, useThemeCtx } from '@retail-ui/theme'
import clsx from 'clsx'
import * as React from 'react'

export type BadgeVariant = keyof Theme['BadgeStyles']['variant']
export type BadgeColor = keyof Theme['BadgeStyles']['variant']['solid']

export type BadgeProps = React.PropsWithChildren<{
  variant?: BadgeVariant
  color?: BadgeColor
  hasClose?: boolean
  className?: string
  onClose?: () => void
}>
type Ref = HTMLDivElement

// TODO: hasClose icon size
export const Badge = React.forwardRef<Ref, BadgeProps>((props, ref) => {
  const {
    children,
    className,
    variant = 'solid',
    color = 'primary',
    hasClose,
    onClose,
  } = props

  const {
    theme: { BadgeStyles },
  } = useThemeCtx()

  const variantCls = BadgeStyles.variant
  const colorCls = variantCls[variant]

  const cls = clsx(
    BadgeStyles.base,
    BadgeStyles.active,
    BadgeStyles.focus,
    colorCls[color],
    className,
  )

  const btnVariant: Record<BadgeVariant, ButtonIconVariant> = {
    solid: 'default',
    light: 'light',
  }
  const btnCls = BadgeStyles.btn

  return (
    <div ref={ref} className={cls}>
      {children}
      {hasClose && (
        <ButtonIcon
          className={btnCls}
          color={color}
          size="xs"
          shape="rounded"
          icon={XSolid}
          variant={btnVariant[variant]}
          onClick={onClose}
        />
      )}
    </div>
  )
})

Badge.displayName = 'Badge'
