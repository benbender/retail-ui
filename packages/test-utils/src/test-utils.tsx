import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
