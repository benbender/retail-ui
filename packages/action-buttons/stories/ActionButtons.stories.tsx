import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
