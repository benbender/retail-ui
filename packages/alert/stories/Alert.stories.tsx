import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
