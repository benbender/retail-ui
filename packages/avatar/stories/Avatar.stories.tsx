import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
