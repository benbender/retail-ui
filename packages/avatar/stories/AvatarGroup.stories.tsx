import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
