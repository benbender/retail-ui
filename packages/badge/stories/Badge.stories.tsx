import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
