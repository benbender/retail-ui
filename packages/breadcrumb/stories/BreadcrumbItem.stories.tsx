import * as React from 'react'

import { Breadcrumb, BreadcrumbItem, BreadcrumbProps } from '../src'

const meta = {
  title: 'Retail-UI/BreadcrumbItem',
  component: BreadcrumbItem,
}

export default meta

const Template = (args: BreadcrumbProps) => (
  <Breadcrumb data-testid={meta.title}>
    <BreadcrumbItem {...args}>
      <a href="#">Dashboard</a>
    </BreadcrumbItem>
    <BreadcrumbItem {...args}>
      <a href="#">Project</a>
    </BreadcrumbItem>
    <BreadcrumbItem isActive {...args}>
      Export
    </BreadcrumbItem>
  </Breadcrumb>
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Breadcrumb',
}
