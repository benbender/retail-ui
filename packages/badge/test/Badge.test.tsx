import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Badge from '../stories/Badge.stories'

describe('Badge', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Badge.Simple />)
    expect(asFragment()).toMatchSnapshot()
  })
})
