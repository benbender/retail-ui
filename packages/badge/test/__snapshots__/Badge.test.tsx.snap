// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders correctly 1`] = `
<DocumentFragment>
  <div
    class="text-center no-underline font-semibold relative overflow-hidden py-1 px-3 w-auto inline-flex items-center leading-none whitespace-no-wrap align-baseline text-xs rounded-full transition-colors duration-150 ease-in cursor-default active:shadow-outline active:outline-none focus:shadow-outline focus:outline-none bg-purple-600 text-white hover:bg-purple-700 focus:bg-purple-700"
  />
</DocumentFragment>
`;
